### PR TITLE
[FSTORE-391] Make Spark DataFrame Columns Nullable on ingestion

### PR DESCRIPTION
**Pull request by tdoehmen**
This PR adds a routine for ingested spark dataframes that automatically makes all columns nullable (except partition key, primary key, and hudi precimbine key)
- added for Python and Java engine
- added tests

JIRA Issue: FSTORE-391

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java [hunk 1]
```diff
@@ -769,10 +769,23 @@ public <S> List<Feature> parseFeatureGroupSchema(S datasetGeneric,
...
-  public <S> S sanitizeFeatureNames(S datasetGeneric) {
+  public <S> S convertToDefaultDataframe(S datasetGeneric, List<String> nonNullColumns) {
     Dataset<Row> dataset = (Dataset<Row>) datasetGeneric;
-    return (S) dataset.select(Arrays.asList(dataset.columns()).stream().map(f -> col(f).alias(f.toLowerCase())).toArray(
-        Column[]::new));
+    Dataset<Row> sanitizedNamesDataset = dataset.select(Arrays.asList(dataset.columns()).stream().map(f ->
+            col(f).alias(f.toLowerCase())).toArray(Column[]::new));
+
+    if (nonNullColumns != null) {
+      Dataset<Row> datasetCopy = sanitizedNamesDataset.as("df1");
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java:778]: not sure why this additional copy is needed?

File: java/src/main/java/com/logicalclocks/hsfs/StreamFeatureGroup.java [hunk 2]
```diff
@@ -428,4 +429,18 @@ public String getEncodedAvroSchema() throws FeatureStoreException, IOException {
   public Schema getDeserializedAvroSchema() throws FeatureStoreException, IOException {
     return utils.getDeserializedAvroSchema(getAvroSchema());
   }
+
+  public List<String> nonNullableColumns() {
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/StreamFeatureGroup.java:433]: Can't we have this in the feature group base, so we don't have to have it twice?

Comment (tdoehmen) [reply java/src/main/java/com/logicalclocks/hsfs/StreamFeatureGroup.java:433]: We don't have the required properties there unfortunately.

File: java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java (hunk 1 above)
Comment (tdoehmen) [reply java/src/main/java/com/logicalclocks/hsfs/engine/SparkEngine.java:778]: Indeed not needed

File: java/src/main/java/com/logicalclocks/hsfs/StreamFeatureGroup.java (hunk 2 above)
Comment (moritzmeister) [reply java/src/main/java/com/logicalclocks/hsfs/StreamFeatureGroup.java:433]: ah right, partition and pre-combine.

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupEngine.java [hunk 3]
```diff
@@ -80,7 +80,7 @@ public void insert(FeatureGroup featureGroup, Dataset<Row> featureData, Storage
                      String hudiPrecombineKey, Map<String, String> writeOptions)
       throws FeatureStoreException, IOException, ParseException {
     Integer validationId = null;
-
+
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureGroupEngine.java:83]:
```suggestion

```

File: java/src/test/java/com/logicalclocks/hsfs/TestFeatureGroup.java [hunk 4]
```diff
@@ -118,4 +118,5 @@ public void testFeatureGroupPartitionPrecombineKeys() {
     Assertions.assertEquals(precombineException.getMessage(),
         "Provided Hudi precombine key hudiPrecombineKey doesn't exist in feature dataframe");
   }
+
```
Comment (moritzmeister) [inline java/src/test/java/com/logicalclocks/hsfs/TestFeatureGroup.java:121]:
```suggestion
```

File: java/src/test/java/com/logicalclocks/hsfs/engine/TestSparkEngine.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        StructType expected = new StructType();
+        expected = expected.add("a", DataTypes.StringType, true);
+        expected = expected.add("b", DataTypes.StringType, true);
+        expected = expected.add("c", DataTypes.StringType, true);
+        expected = expected.add("d", DataTypes.StringType, true);
+        ArrayList<StructField> expectedJava = new ArrayList<>(JavaConverters.asJavaCollection(expected.toSeq()));
+
+        ArrayList<StructField> dfConvertedJava = new ArrayList<>(JavaConverters.asJavaCollection(dfConverted.schema().toSeq()));
+        // Assert
```
Comment (moritzmeister) [inline java/src/test/java/com/logicalclocks/hsfs/engine/TestSparkEngine.java:48]: Do we need an assertion to make sure the original schema is unchanged?

Comment (tdoehmen) [reply java/src/test/java/com/logicalclocks/hsfs/engine/TestSparkEngine.java:48]: added

**Review by moritzmeister — APPROVED** [review]

**Pull request merged**